x/cosmwasmpool/pool_module_test.go

```diff
@@ -0,0 +1,338 @@
+
+			expectedErrorMessage: types.InvalidPoolTypeError{
+				ActualPool: &clmodel.Pool{},
+			}.Error(),
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		s.Run(name, func() {
+			s.SetupTest()
+
+			cosmwasmPoolKeeper := s.App.CosmwasmPoolKeeper
+
+			// fund pool joiner
+			s.FundAcc(s.TestAccs[0], tc.initialCoins)
+
+			// get initial denom from coins specified in the test case
+			initialDenoms := []string{}
+			for _, coin := range tc.initialCoins {
+				initialDenoms = append(initialDenoms, coin.Denom)
+			}
+
+			// create pool
+			pool := s.PrepareCustomTransmuterPool(s.TestAccs[0], initialDenoms, 1)
+
+			// add liquidity by joining the pool
+			request := transmuter.JoinPoolExecuteMsgRequest{}
+			cosmwasm.MustExecute[transmuter.JoinPoolExecuteMsgRequest, msg.EmptyStruct](s.Ctx, s.App.ContractKeeper, pool.GetContractAddress(), s.TestAccs[0], tc.initialCoins, request)
+
+			originalPoolBalances := s.App.BankKeeper.GetAllBalances(s.Ctx, sdk.AccAddress(pool.GetContractAddress()))
+
+			var poolIn poolmanagertypes.PoolI = pool
+			if tc.isInvalidPool {
+				poolIn = s.PrepareConcentratedPool()
+			}
+
+			// system under test non-mutative.
+			actualCalcTokenOut, err := cosmwasmPoolKeeper.CalcOutAmtGivenIn(s.Ctx, poolIn, tc.tokenIn, tc.tokenOutDenom, tc.swapFee)
+			if tc.expectedErrorMessage != "" {
+				s.Require().Error(err)
+				s.Require().ErrorContains(err, tc.expectedErrorMessage)
+			} else {
+				s.Require().NoError(err)
+				s.Require().Equal(tc.expectedTokenOut, actualCalcTokenOut)
+			}
+
+			// Assert that pool balances are unchanged
+			afterCalcPoolBalances := s.App.BankKeeper.GetAllBalances(s.Ctx, sdk.AccAddress(pool.GetContractAddress()))
+
+			s.Require().Equal(originalPoolBalances.String(), afterCalcPoolBalances.String())
+
+			swapper := s.TestAccs[1]
+			// fund swapper
+			s.FundAcc(swapper, sdk.NewCoins(tc.tokenIn))
+
+			// system under test non-mutative.
+			actualSwapTokenOut, err := cosmwasmPoolKeeper.SwapExactAmountIn(s.Ctx, swapper, poolIn, tc.tokenIn, tc.tokenOutDenom, tc.tokenOutMinAmount, tc.swapFee)
+			if tc.expectedErrorMessage != "" {
+				s.Require().Error(err)
+				s.Require().ErrorContains(err, tc.expectedErrorMessage)
+
+				// TODO: check that pool balances are unchanged
+
+				return
+			}
+
+			s.Require().NoError(err)
+			s.Require().Equal(tc.expectedTokenOut.Amount, actualSwapTokenOut)
+
+			// Assert that pool and swapper balances are changes
+			// TODO / Question for Boss: this check is failint because the originalPoolBalance is empty.
+			// I would expect it to equal to coins that were added to the pool in the beginning of the test
+			// via the join pool message (tc.initialCoins).
+			expectedPoolBalances := originalPoolBalances.Add(tc.tokenIn).Sub(sdk.NewCoins(tc.expectedTokenOut))
+			afterSwapPoolBalances := s.App.BankKeeper.GetAllBalances(s.Ctx, sdk.AccAddress(pool.GetContractAddress()))
+
+			s.Require().Equal(expectedPoolBalances.String(), afterSwapPoolBalances.String())
```

Comment:
@iboss-ptk it seems to me that pool's balances are not updated upon joining the pool. Should this be handles inside a contract? That is when a user executes join message via contract, the pool's balance is updated with the joined tokens?